components/net_traits/pub_domains.rs

```diff
@@ -146,6 +146,12 @@ pub fn is_reg_domain(domain: &str) -> bool {
 /// Returns None if the URL has no host name.
 /// Returns the registered suffix for the host name if it is a domain.
 /// Leaves the host name alone if it is an IP address.
-pub fn reg_host<'a>(url: &'a ServoUrl) -> Option<&'a str> {
-    url.domain().map(reg_suffix).or(url.host_str())
+pub fn reg_host(url: &ServoUrl) -> Option<Host> {
+    match url.origin() {
+        ImmutableOrigin::Tuple(_, host, _) => match host {
+            Host::Domain(domain) => Some(Host::Domain(String::from(reg_suffix(&*domain)))),
+            _ => Some(host),
+        },
+        ImmutableOrigin::Opaque(_) => None,
+    }
```

Comment:
Or a mix of the two.

```rust
Nit: you can merge the two arms above together AFAICT.

```rust
match url.origin() {
    ImmutableOrigin::Tuple(_, Host::Domain(domain), _) => {
        Some(Host::Domain(String::from(reg_suffix(&*domain))))
    },
    ImmutableOrigin::Tuple(_, ip, _) => Some(ip),
    ImmutableOrigin::Opaque(_) => None,
}
```